skglm/solvers/gram.py

```diff
@@ -0,0 +1,235 @@
+import numpy as np
+from numba import njit
+from numpy.linalg import norm
+from celer.homotopy import _grp_converter
+
+from skglm.utils import BST, ST, BST_vec, ST_vec
+
+
+@njit
+def primal(alpha, r, w, weights):
+    p_obj = (r @ r) / (2 * len(r))
+    return p_obj + alpha * np.sum(np.abs(w * weights))
+
+
+@njit
+def primal_grp(alpha, norm_r2, r, w, grp_ptr, grp_indices, weights):
+    p_obj = norm_r2 / (2 * len(r))
+    for g in range(len(grp_ptr) - 1):
+        w_g = w[grp_indices[grp_ptr[g]:grp_ptr[g + 1]]]
+        p_obj += alpha * norm(w_g * weights[g], ord=2)
+    return p_obj
+
+
+@njit
+def dual(alpha, norm_y2, theta, y):
+    d_obj = - np.sum((y / (alpha * len(y)) - theta) ** 2)
+    d_obj *= 0.5 * alpha ** 2 * len(y)
```

Comment:
@mathurinm, 
shouldn't use the same convention as in celer
``theta = y - Xw`` instead of ``theta = (y - Xw) / alpha``

Reply:
Good point ! This PR will not be merged as is, we don't use the gap in skglm; it's just to do a bench with the same stopping criterion as in celer. So not a problem here